[WIP] Integration test suite harness for Native Image, feat. Quarkus, Micronaut, Helidon and more

Karm opened the pull request:
There will be a debate as to whether this belongs to the repo itself or
to a separate new repository.

It was tested on Linux. Windows runs are WIP at the moment, but it was developed with Windows on mind.

See [README.md](https://github.com/Karm/graal/blob/e68905f3a7e5b5b3708e66ee3107cdec0dd97c8c/integration-tests/README.md)

zakkak commented on integration-tests/ACKNOWLEDGEMENTS.md:
````diff
@@ -0,0 +1,21 @@
+Test application ```quarkus-full-microprofile``` is a modified version of code 
+originally generated by [https://start.microprofile.io/](https://start.microprofile.io/). 
+Some utility methods were adopted from [https://github.com/eclipse/microprofile-starter/](https://github.com/eclipse/microprofile-starter/) 
+and modified to suit the purpose of this TS. The adopted work was authored 
````
```suggestion
and modified to suit the purpose of this test suite (TS). The adopted work was authored 
```

Karm replied:
Ack

zakkak commented on integration-tests/ACKNOWLEDGEMENTS.md:
```diff
@@ -0,0 +1,21 @@
+Helidon example app was adapted from a Helidon SE QuickStart archetype. See [QuickStart](https://helidon.io/docs/latest/#/se/guides/02_quickstart).
+
+## Windows console and CTRL+C
+The method of handling the CTRL+C was adopted by Michal Karm Babacek <karm@redhat.com> from hints by [Stanislav](http://stackoverflow.com/a/15281070/32453) and [rogerdpack](https://stackoverflow.com/users/32453/rogerdpack).
```
It might be better to link to the corresponding answer/comment instead of linking to rogerdpack's profile (like you already do for Stanislav's answer.

Karm replied:
Ack.

zakkak commented on integration-tests/README.md:
````diff
@@ -0,0 +1,121 @@
+
+## Prerequisites
+
+The TS expects you run Java 11+ and have ```ps``` program available on your Linux/Mac
````
```suggestion
The test suite (TS) expects you run Java 11+ and have ```ps``` program available on your Linux/Mac
```

Karm replied:
Ack.

zakkak commented on integration-tests/README.md:
````diff
@@ -0,0 +1,121 @@
+
+The basic usage that runs all tests:
+```
+mvn clean verify -Ptestsuite
````
Out of curiosity, why not just `mvn clean verify`?

Karm replied:
@zakkak The test apps are modules and they are included in the parent by default, so as you have them imported in your IDE as a one project without any hassle. When you run the test suite, you don't want to build those apps too; you just want to run the TS. The TS builds the apps at its own discretion at the right time.  Hence -Ptestsuite says "Just run the TS, don't build apps/* modules..."

If your Maven-fu could bring a better way of doing this without hampering the IDE experience, I am all ears...

zakkak replied:
Thanks for the clarification @Karm .

If I understand correctly both `mvn clean verify -Ptestsuite` and `mvn clean verify` will have the same **result**, but `mvn clean verify -Ptestsuite` might be faster in some cases.

The difference is that `mvn clean verify` will decent in the apps and do `mvn clean verify` there as well, while `mvn clean verify -Ptestsuite` will only run in testsuite (it is actually the same with `mvn clean verify -pl testsuite`) and the test suite will implicitly invoke `mvn ... package ...` in the apps to build them as needed.

My Maven-fu ( :) ) approach would probably use `io.quarkus:quarkus-maven-plugin` for creating the native images in the apps pom files, and avoid invoking mvn from the testsuite harness.

The entry would look like this (copied from a quarkus integration test):
```xml
                    <plugin>
                        <groupId>io.quarkus</groupId>
                        <artifactId>quarkus-maven-plugin</artifactId>
                        <executions>
                            <execution>
                                <id>native-image</id>
                                <goals>
                                    <goal>native-image</goal>
                                </goals>
                                <configuration>
                                    <cleanupServer>true</cleanupServer>
                                    <enableHttpUrlHandler>true</enableHttpUrlHandler>
                                    <graalvmHome>${graalvmHome}</graalvmHome>
                                </configuration>
                            </execution>
                        </executions>
                    </plugin>
```

Not sure it's worth the extra effort though. Just sharing my thoughts.

Karm replied:
@zakkak I don't understand how this helps Helidon or Micronaut  (and eventually other modules)...? 
I might be missing the point of what is bothering you in the first place. Is it that the default profile is suited for IDE and not for execution? Why would it be a problem?

zakkak replied:
No problem here, just discussing and sharing my thoughts, feel free to ignore me :) Sorry for not making this clear.

With this plugin you should be able to create the native-image directly through maven (for any framework), without the need to invoke `mvn` and `native-image` from the test harness. In theory this could make configuration easier as well (e.g. for using a different `native-image` executable.

zakkak commented on integration-tests/ACKNOWLEDGEMENTS.md:
```diff
@@ -0,0 +1,21 @@
+# Acknowledgements
```
You could integrate this in README.md

Karm replied:
@zakkak Dunno...Would you find it better? I feel like the Acknowledgements belong more to the License / Copyright notice files than to the user manual.  I don't have a strong opinion on it though...

zakkak replied:
Me neither, I have seen it both as a separate file and as a section in Readme.

zakkak commented on integration-tests/README.md:
````diff
@@ -0,0 +1,121 @@
+```
+org.opentest4j.AssertionFailedError: 
+Application QUARKUS_FULL_MICROPROFILE took 27 ms to get the first OK request, 
+which is over 5 ms threshold. ==> expected: <true> but was: <false>
````
The `expected: <true> but was: <false>` is slightly confusing, I guess it refers to `27 <= 5`?
Unfortunately I can only think about adding the actual check just before `==>`,  to improve this.

Karm replied:
@zakkak junit.jupiter.api doesn't give me GreaterThan / LessThan IMHO... So it boils down to this:

```
assertTrue(timeToFirstOKRequest <= timeToFirstOKRequestThresholdMs,
               "Application " + app + " took " + timeToFirstOKRequest
                + " ms to get the first OK request, which is over " +
                timeToFirstOKRequestThresholdMs + " ms threshold.");
```

zakkak replied:
Thought so. What I am suggesting is:
```java
assertTrue(timeToFirstOKRequest <= timeToFirstOKRequestThresholdMs,
               "Application " + app + " took " + timeToFirstOKRequest
                + " ms to get the first OK request, which is over " +
                timeToFirstOKRequestThresholdMs + " ms threshold. (" + app + " <= " + timeToFirstOKRequest + ")");
```
But that's probably more confusing.

There is a better approach discussed [here](https://stackoverflow.com/a/18766311/1119431) but I am probably nit-picking.

zakkak commented on integration-tests/README.md:
````diff
@@ -0,0 +1,121 @@
+    [WARNING] Discovered module-info.class. Shading will break its strong encapsulation.
+```
+
+Happy testing!
````
```suggestion
**Happy testing!**
```
:)

zakkak commented on integration-tests/testsuite/src/it/java/org/graalvm/tests/integration/utils/BuildAndRunCmds.java:
```diff
@@ -0,0 +1,56 @@
+public enum BuildAndRunCmds {
+    // Note that at least 2 command are expected. One or more to build. The last one to run the app.
+    QUARKUS_FULL_MICROPROFILE(new String[][]{
+            new String[]{"mvn", "clean", "compile", "package", "-Pnative"},
```
How does `mvn` pick `native-image` in this case?

Karm replied:
@zakkak It differs runtime to runtime I am afraid. Some just expect it on PATH. Quarkus looks for GRAALVM_HOME too.

The TS as it is right now is oblivious to native-image location; it just expects it on your PATH.  Maybe I should enhance the README...

zakkak replied:
If it's going to be part of mandrel/graal rerpo one would expect it to test the latest/current build I guess, so that could be confusing.

I agree that it needs to be documented in the README, and possibly extended at a later stage to be configurable without having to change `PATH`.

zakkak commented on integration-tests/testsuite/src/it/java/org/graalvm/tests/integration/utils/BuildAndRunCmds.java:
```diff
@@ -0,0 +1,56 @@
+    }),
+    MICRONAUT_HELLOWORLD(new String[][]{
+            new String[]{"mvn", "package"},
+            new String[]{"native-image", "-jar", "target/helloworld.jar", "target/helloWorld"},
```
I think that we should also be able to control which `native-image` to use.

Karm replied:
@zakkak I see. That is not the case in this implementation, it just picks what you've got on your PATH.   I wanted it to start simple...

zakkak replied:
Same as in https://github.com/graalvm/mandrel/pull/83#discussion_r459993937

zakkak commented on integration-tests/testsuite/src/it/java/org/graalvm/tests/integration/utils/WhitelistLogLines.java:
```diff
@@ -0,0 +1,61 @@
+    }),
+
+    HELIDON_QUICKSTART_SE(new Pattern[]{
+            // Unused argument on new Graal; Quarkus uses it for backward compatibility.
```
If true:

```suggestion
            // Unused argument on new Graal; Helidon uses it for backward compatibility.
```

Karm replied:
Ack.

zakkak commented on integration-tests/apps/quarkus-full-microprofile/src/main/resources/privateKey.pem:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Just to be sure. This is a dummy right?

Karm replied:
It is. Generated just for this TS.

zakkak commented on integration-tests/testsuite/src/it/resources/README.md:
```diff
@@ -0,0 +1,38 @@
+# CtrlC.exe
```
I think there should be a link from the main README.md to this file.

Karm replied:
@zakkak Why do you think it is important for the main README? The only reason that comes to my mind would be to explain as early as possible that that is a strange .exe file in the repo so as people don't freak out about it prematurely.   
Otherwise it is an implementation detail (although it took me quite some time to figure out back in the day :see_no_evil: )

zakkak replied:
Yes, if I am going to run this on windows I want to know that a strange binary that I didn't build is going to run on my machine. Ideally I would prefer to build it from source code, not sure how hard that would be though.

Karm replied:
@zakkak I built that binary (I did not download a random binary and put it in the repo :) ) and it is trivial. Given the fact that running this TS on WIndows without VS doesn't make sense, people are going to have the toolchain installed anyway. I can build it on the fly.     
It would just need to be done at the right time. Lemme think about it. Having any extra scripts defeats the purpose of having Maven there in the first place, so I would probably create some aux module to be built before testsuite module starts....

Karm replied:
Anyway, I expect having more small native utilities there over the time...

zakkak replied:
> @zakkak I built that binary (I did not download a random binary and put it in the repo :) ) 

I agree, but anyone else who will download the testsuite and run it will essential _download a random binary and run it_ so they might not feel good about it.

I see how solving this is not that trivial, especially with proper maven integration. What if you build it through the test harness (like you invoke native-image)?

Karm commented:
TODO: WIndows

It needs some Whitelisting and some investigation:

```
[ERROR] micronautHelloWorld{TestInfo}  Time elapsed: 213.223 s  <<< FAILURE!
org.opentest4j.AssertionFailedError:
build-and-run.log log should not contain error or warning lines that are not whitelisted. See testsuite\target\archived-logs\org.graalvm.tests.integration.RuntimesSmokeTest\micronautHelloWorld\build-and-run.log and check these offending lines:
14:20:40,127 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - Failed to create WindowsAnsiOutputStream. Falling back on the default stream. ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type org.fusesource.jansi.WindowsAnsiOutputStream ==> expected: <true> but was: <false>
        at org.graalvm.tests.integration.RuntimesSmokeTest.testRuntime(RuntimesSmokeTest.java:118)
        at org.graalvm.tests.integration.RuntimesSmokeTest.micronautHelloWorld(RuntimesSmokeTest.java:153)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   AppReproducersTest.randomNumbersReinit:121 There should have been 4 distinct lines in the log,showing 2 different pseudorandom sequences. The fact that there are less than 4 means the native imagewas not properly re-seeded. See https://github.com/oracle/graal/issues/2265. ==> expected: <4> but was: <0>
[ERROR]   RuntimesSmokeTest.helidonQuickStart:159->testRuntime:100 Timeout 10s was reached. Empty webpage does not contain string: `Hello World!' ==> expected: <true> but was: <false>
[ERROR]   RuntimesSmokeTest.micronautHelloWorld:153->testRuntime:118 build-and-run.log log should not contain error or warning lines that are not whitelisted. See testsuite\target\archived-logs\org.graalvm.tests.integration.RuntimesSmokeTest\micronautHelloWorld\build-and-run.log and check these offending lines:
14:20:40,127 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - Failed to create WindowsAnsiOutputStream. Falling back on the default stream. ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type org.fusesource.jansi.WindowsAnsiOutputStream ==> expected: <true> but was: <false>
[ERROR]   RuntimesSmokeTest.quarkusFullMicroProfile:147->testRuntime:118 build-and-run.log log should not contain error or warning lines that are not whitelisted. See testsuite\target\archived-logs\org.graalvm.tests.integration.RuntimesSmokeTest\quarkusFullMicroProfile\build-and-run.log and check these offending lines:
[WARNING] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Uber JAR strategy is used for native image source JAR generation on Windows. This is done for the time being to work around a current GraalVM limitation on Windows concerning the maximum command length (see https://github.com/oracle/graal/issues/2387). ==> expected: <true> but was: <false>
[INFO]
[ERROR] Tests run: 4, Failures: 4, Errors: 0, Skipped: 0
```

Karm commented:
Helidon failure: https://github.com/oracle/helidon/issues/2230

Karm commented:
@zakkak  I addressed the concerns I believe and I successfully ported it to Windows. I will carry on with it here:

https://github.com/Karm/mandrel-integration-tests

And then propose it on graalvm-dev@oss.oracle.com for inclusion in either https://github.com/oracle/graal as a subdirectory or perhaps better as https://github.com/mandrel/mandrel-integration-tests.
